main.go

```diff
@@ -154,3 +189,247 @@ func gatekeeperNamespace(platformInfo platform.PlatformInfo) (string, error) {
+	if err != nil {
+		setupLog.Error(err, "Error to setup watcher")
+		ctxCancel()
+		return
+	}
+
+	for {
+		select {
+		case result, ok := <-watcher.ResultChan():
+			// the channel got closed, so we need to restart
+			if !ok {
+				watcher, err = newWatcher(mainCtx, dynamicClient)
+				if err != nil {
+					setupLog.Error(err, "Error to setup watcher")
+					ctxCancel()
+					return
+				}
+
+				continue
+			}
+
+			switch result.Type { //nolint:exhaustive
+			case watch.Modified, watch.Added:
+				if isInstalled {
+					continue
+				}
+
+				// Check constraintpodstatuses crd available
+				if !checkCPScontrollerAvailable(mainCtx, result, dynamicClient) {
+					continue
+				}
+
+				isInstalled = true
+
+				go func() {
+					// ConstraintPodStatus manager
+					cpsMgr, err := ctrl.NewManager(cfg, ctrl.Options{
+						Scheme: scheme,
+						Metrics: server.Options{
+							BindAddress: ":8083",
+						},
+						HealthProbeBindAddress: ":8084",
+						LeaderElection:         enableLeaderElection,
+						LeaderElectionID:       "5ff985ccc.constraintstatuspod.gatekeeper.sh",
+						Cache: cacheRuntime.Options{
+							ByObject: map[client.Object]cacheRuntime.ByObject{
+								&v1beta1.ConstraintPodStatus{}: {
+									Transform: func(obj interface{}) (interface{}, error) {
+										constraintStatus := obj.(*v1beta1.ConstraintPodStatus)
+										// Only cache fields that are utilized by the controllers.
+										guttedObj := &v1beta1.ConstraintPodStatus{
+											TypeMeta: constraintStatus.TypeMeta,
+											ObjectMeta: metav1.ObjectMeta{
+												Name:      constraintStatus.Name,
+												Labels:    constraintStatus.Labels,
+												Namespace: constraintStatus.Namespace,
+											},
+											Status: v1beta1.ConstraintPodStatusStatus{
+												ObservedGeneration: constraintStatus.Status.ObservedGeneration,
+												Operations:         constraintStatus.Status.Operations,
+											},
+										}
+										return guttedObj, nil
+									},
+								},
+							},
+						},
+					})
+					if err != nil {
+						setupLog.Error(err, "unable to setup ConstraintPodStatus manager")
+					}
+					// Start with new context since previous context was cancelled
+					err = addConstraintStatusManager(ctx, cpsMgr, dynamicClient, namespace)
+					if err != nil {
+						setupLog.Error(err, "unable to start ConstraintPodStatus manager")
+					}
+				}()
+
+			case watch.Deleted:
+				ctxCancel()
+				isInstalled = false
+				setupLog.Info("Gatekeeper deleted. Close ConstraintPodStatus manager")
+				// Override ctx, ctxCancel
+				ctx, ctxCancel = context.WithCancel(mainCtx)
+			}
+		case <-mainCtx.Done():
+			// When parents ctx close, then child should be closed
+			ctxCancel()
+			return
+		}
+	}
+}
+
+func newWatcher(mainCtx context.Context,
+	dynamicClient *dynamic.DynamicClient,
+) (*toolsWatch.RetryWatcher, error) {
+	timeout := int64(30)
+	gatekeeperGVR := schema.GroupVersionResource{
```

Comment:
I have concerns with this approach because if the CRD takes longer than 30 seconds, then the manager is not started. Then you have to wait for the Gatekeeper object to change to trigger the code path of checking the CRD.

It occurred to me that `GatekeeperReconciler` already watches the `Gatekeeper` object. Why not start the manager in there? After it install Gatekeeper, it can check if the CRD is ready. If not, it requeues the request with a `requestAfter` delay of 5 seconds or so. You would need to change the predicate to allow reconciles to deletes to stop the manager though.

Reply:
CRD is not ready we cannot start `crtl.manager cache`

Reply:
good idea Let me try